src/loki/reach.cc

```diff
@@ -110,17 +131,117 @@ directed_reach SimpleReach(const DirectedEdge* edge,
         continue;
       const auto* node = tile->node(edge.endnode());
       const auto* opp_edge = tile->directededge(node->edge_index() + edge.opp_index());
+
+      // NOTE: we can go through the end of the restriction because only the start would mark a
+      // potential stopping point (maybe a path followed the restriction). We also have to stop
+      // at the start of a simple restriction because it could have been on our path
```

Comment:
and similarly we have to stop the expansion if we see the start of any kind of restriction in case the path would have been over the whole restriction.